### PR TITLE
Make changes for the retriever scripts being shipped to a separate repository

### PR DESCRIPTION
**Pull request by harshitbansal05**

### CODE REVIEW

File: hooks/pre-commit [hunk 1]
```diff
@@ -47,7 +47,7 @@ def get_script_version(file_name):
 
 # check and get script directory changes
 subprocess.check_output(
-    'git fetch https://github.com/weecology/retriever.git master', shell=True)
+    'git fetch https://github.com/harshitbansal05/retriever.git development', shell=True)
```
Comment (henrykironde) [inline hooks/pre-commit]: Do we want to keep this here `harshitbansal05`

File: hooks/pre-commit [hunk 2]
```diff
@@ -63,8 +63,12 @@ modified_staged_scripts = to_string(subprocess.check_output(
 added_deleted_staged_scripts = to_string(subprocess.check_output(
     "git diff --cached --name-only --diff-filter=AmD scripts", shell=True))
 
+if added_deleted_staged_scripts != "":
+	print("\nThe legacy scripts of Retriever cannot be added to or removed. To add scripts,\
+	use the retriever-recipes repository.")
+	exit(1)
```
Comment (henrykironde) [inline hooks/pre-commit]: Avoid using exit, instead return would be better. exit(1) will cause the tests to fail in later edge cases especially in the Retriever.Julia package and in the rdataretriever package.

File: hooks/pre-commit [hunk 3]
```diff
@@ -88,10 +92,9 @@ if modified_staged_scripts != "":
...
-# only runs version.py if we have modified scripts which have had their versions updated,
-# or we have no modified scripts but either deleted or added scripts.
-if (modified_staged_scripts and versions_updated) or (added_deleted_staged_scripts != ""):
+# only runs version.py if we have modified scripts which have had their versions updated.
+if (modified_staged_scripts and versions_updated):
     print("Updating version.txt")
     subprocess.call(['python', 'version.py'])
     print("version.txt added to commit")
-    subprocess.call(['git', 'add', 'version.txt'])
+    subprocess.call(['git', 'add', 'version.txt'])
```
Comment (henrykironde) [inline hooks/pre-commit]: Line end

File: hooks/pre-commit (hunk 1 above)
Comment (harshitbansal05) [reply hooks/pre-commit]: Right now, I kept this only for testing. Later, we can change it to retriever repository.

Comment (zhangcandrew) [reply hooks/pre-commit]: I think it's better if we change it to the retriever-recipes repository right now instead of your temporary git. We can wait for [retriever-recipes PR 2](https://github.com/weecology/retriever-recipes/pull/2) to go in first before merging this in.

File: retriever/lib/get_opts.py [hunk 4]
```diff
@@ -52,9 +49,6 @@
 default_parser = subparsers.add_parser('defaults', help='displays default options')
 update_parser = subparsers.add_parser('update', help='download updated versions of scripts')
 new_parser = subparsers.add_parser('new', help='create a new sample retriever script')
```
Comment (zhangcandrew) [inline retriever/lib/get_opts.py]: Is it possible to make this new parser create scripts within the scripts folder

File: retriever/lib/repository.py [hunk 5]
```diff
@@ -69,9 +70,14 @@ def check_for_updates():
...
-                                              os.path.normpath(os.path.join(SCRIPT_WRITE_PATH, script_name)))
+                                              os.path.normpath(os.path.join(SCRIPT_WRITE_PATH, script_name)),
+                                              repo)
                 except Exception as e:
                     print(e)
                     pass
     except:
         raise
+
+    if repo == RETRIEVER_REPOSITORY:
```
Comment (zhangcandrew) [inline retriever/lib/repository.py]: Can you run me through your logic for checking if the repo is a scripts repo vs just the normal retriever repo?

File: retriever/lib/scripts.py [hunk 6]
```diff
@@ -161,6 +163,61 @@ def to_str(object, object_encoding=sys.stdout, object_decoder=ENCODING):
...
+        if json_file_encoding:
+            file_obj = open_fr(json_file, encoding=json_file_encoding)
+        else:
+            file_obj = open_fr(json_file)
+        json_object = json.load(file_obj)
+        file_obj.close()
+    except ValueError:
+        pass
+
+    if type(json_object) is dict and "version" in json_object.keys():
```
Comment (zhangcandrew) [inline retriever/lib/scripts.py]: perhaps: 

`return json_object.get("version", None)`?

**Comment by zhangcandrew** [comment]
Thanks @harshitbansal05, It's looking good, I'm going to Pull it down and test some parts out when I get the chance!

File: hooks/pre-commit (hunk 1 above)
Comment (harshitbansal05) [reply hooks/pre-commit]: Ok, I would change it as soon as recipes pr is merged.

File: hooks/pre-commit (hunk 2 above)
Comment (harshitbansal05) [reply hooks/pre-commit]: Would do so.

File: retriever/lib/get_opts.py (hunk 4 above)
Comment (harshitbansal05) [reply retriever/lib/get_opts.py]: We can so it, should I open a separate pr for it or do it in this?

File: hooks/pre-commit (hunk 3 above)
Comment (harshitbansal05) [reply hooks/pre-commit]: Noted.

File: retriever/lib/repository.py (hunk 5 above)
Comment (harshitbansal05) [reply retriever/lib/repository.py]: @zhangcandrew in the method `check_for_updates`, it is done from both the repositories, retriever (for the five legacy scripts) and recipes (for the rest of the scripts). I just used the above  logic to call the same method with different repositories (to update from both of them).

File: retriever/lib/scripts.py (hunk 6 above)
Comment (harshitbansal05) [reply retriever/lib/scripts.py]: Noted.

**Comment by harshitbansal05** [comment]
@henrykironde is it fine now?

File: retriever/lib/get_opts.py (hunk 4 above)
Comment (zhangcandrew) [reply retriever/lib/get_opts.py]: Let's open a separate PR for this

File: retriever/lib/repository.py (hunk 5 above)
Comment (zhangcandrew) [reply retriever/lib/repository.py]: Ok I see, so this is a placeholder for now, got it!

File: setup.py [hunk 7]
```diff
@@ -139,3 +135,7 @@ def read(*names, **kwargs):
     compile()
 except:
     pass
+
+if os.path.exists(".git/hooks"):  # check if we are in git repo
```
Comment (zhangcandrew) [inline setup.py]: Is there a reason for moving this to the bottom?

File: test/test_modified_scripts.py [hunk 8]
```diff
@@ -37,10 +32,9 @@ def setup_module():
 def get_modified_scripts():
     """Get modified script list, using version.txt in repo and master upstream"""
 
-    os.chdir(retriever_root_dir)
     modified_list = []
-    version_file = requests.get("https://raw.githubusercontent.com/weecology/retriever/master/version.txt")
-    local_repo_scripts = get_script_version()  # local repo versions
+    version_file = requests.get("https://raw.githubusercontent.com/harshitbansal05/retriever/development/version.txt")
```
Comment (zhangcandrew) [inline test/test_modified_scripts.py]: same as above, let's make this the retriever-recipes-dev branch that was just merged in

File: test/test_modified_scripts.py [hunk 9]
```diff
@@ -52,11 +46,8 @@ def get_modified_scripts():
         local_script, local_version = item.lower().split(",")
         # check for new scripts or a change in versions for present scripts
         # repo script versions compared with upstream.
-        if local_script not in upstream_versions.keys():
+        if LooseVersion(local_version) != upstream_versions[local_script]:
```
Comment (zhangcandrew) [inline test/test_modified_scripts.py]: I think this should be greater than since if we have an outdated script locally, it probably wasn't modified and we probably don't care to test it. Maybe @henrykironde has more to say on this though

File: test/test_modified_scripts.py [hunk 10]
```diff
@@ -108,13 +95,13 @@ def install_modified():
                   'table_name': '{db}.{table}'},
 
         "xml": {'engine': 'xml',
-                'table_name': 'output_file_{table}.xml'},
+                'table_name': '{db}_{table}.xml'},
```
Comment (zhangcandrew) [inline test/test_modified_scripts.py]: @henrykironde is this formatting ok? I know we had trouble in the past with some output names but I'm not knowledgeable on the current status of how this works.

File: hooks/pre-commit (hunk 1 above)
Comment (zhangcandrew) [reply hooks/pre-commit]: alright, recipes pr should be merged

**Comment by zhangcandrew** [comment]
I think it's coming together. One thing I forgot to mention, can you output a line in the retriever ls function that says something like 

"For a full list of all script available to download, go to: [retriever-recipes-link] here". 

Also, on update, is it possible to note a distinction of which scripts are donwloading? (like downloading default scripts, downloading upstream scripts) something like that. It currently shows like this: 
image

File: hooks/pre-commit (hunk 1 above)
Comment (harshitbansal05) [reply hooks/pre-commit]: @zhangcandrew the above url refers to the retriever repository, which has not been merged. I have updated the url for the recipes scripts repository.

File: setup.py (hunk 7 above)
Comment (harshitbansal05) [reply setup.py]: No, I don't know how. I would correct it.

File: test/test_modified_scripts.py (hunk 8 above)
Comment (harshitbansal05) [reply test/test_modified_scripts.py]: Again, I have updated the url for the recipes repository. I didn't update this one so the pr could be tested prior to merging it in the retriever repo.

File: test/test_modified_scripts.py (hunk 9 above)
Comment (harshitbansal05) [reply test/test_modified_scripts.py]: I agree with you @zhangcandrew.

File: test/test_modified_scripts.py (hunk 10 above)
Comment (henrykironde) [reply test/test_modified_scripts.py]: You should checkout the format from https://github.com/weecology/retriever/blob/961974cb18940e26d5dd86f5f04ed157d08b3c7f/test/test_integration.py#L546

File: setup.py (hunk 7 above)
Comment (zhangcandrew) [reply setup.py]: I think you can try `git checkout HEAD -- setup.py` to reset it to head.

File: test/test_modified_scripts.py (hunk 8 above)
Comment (zhangcandrew) [reply test/test_modified_scripts.py]: sure that makes sense, but is there a reason why you didn't use the official weecology retriever-recipes version.txt? (https://raw.githubusercontent.com/weecology/retriever-recipes/retriever-recipes-dev/version.txt)

image

File: hooks/pre-commit (hunk 1 above)
Comment (zhangcandrew) [reply hooks/pre-commit]: ahh ok I see what you mean by this then. I see you have 6 commits, let's split this PR into two PR's one with just the changes up to 78f5d64eeb09a9194dc5a75cadba3d029a9ee702, and we can have another with commits after. That way we can get the first half in and then commit the second half successfully without having to reference an intermediary repo like this.

File: test/test_modified_scripts.py (hunk 10 above)
Comment (zhangcandrew) [reply test/test_modified_scripts.py]: ok just double checking, this looks about right

**Comment by zhangcandrew** [comment]
Closing PR in reference to #1345. Feel free to re-open if anyone feels the need
